genienlp/server.py

```diff
@@ -162,46 +162,49 @@ def handle_request(self, request):
-                    self.model, [batch], self.numericalizer, task, self.args, output_predictions_only=True
-                )
-                if sum(self.args.num_outputs) > 1:
+        try:
+            with torch.no_grad():
+                if self.args.calibrator_paths is not None:
+                    output = generate_with_model(
+                        self.model,
+                        [batch],
+                        self.numericalizer,
+                        task,
+                        self.args,
+                        output_predictions_only=True,
+                        confidence_estimators=self.confidence_estimators,
+                    )
                     response = []
-                    for idx, predictions in enumerate(output.predictions):
-                        candidates = []
-                        for cand in predictions:
-                            candidates.append({'answer': cand})
-                        response.append({'candidates': candidates})
+                    if sum(self.args.num_outputs) > 1:
+                        for idx, predictions in enumerate(output.predictions):
+                            candidates = []
+                            for cand in predictions:
+                                candidate = {'answer': cand, 'score': {}}
+                                for e_idx, estimator_scores in enumerate(output.confidence_scores):
+                                    candidate['score'][self.estimator_filenames[e_idx]] = float(estimator_scores[idx])
+                                candidates.append(candidate)
+                            response.append({'candidates': candidates})
+                    else:
+                        for idx, p in enumerate(output.predictions):
+                            instance = {'answer': p[0], 'score': {}}
+                            for e_idx, estimator_scores in enumerate(output.confidence_scores):
+                                instance['score'][self.estimator_filenames[e_idx]] = float(estimator_scores[idx])
+                            response.append(instance)
                 else:
-                    response = [{'answer': p[0]} for p in output.predictions]
+                    output = generate_with_model(
+                        self.model, [batch], self.numericalizer, task, self.args, output_predictions_only=True
+                    )
+                    if sum(self.args.num_outputs) > 1:
+                        response = []
+                        for idx, predictions in enumerate(output.predictions):
+                            candidates = []
+                            for cand in predictions:
+                                candidates.append({'answer': cand})
+                            response.append({'candidates': candidates})
+                    else:
+                        response = [{'answer': p[0]} for p in output.predictions]
+        except RuntimeError:
```

Comment:
How do we check if this is a CUDA error or not?

Reply:
I added a test script, and the exit code 100 was caught.

Reply:
I think the problem is that other non-CUDA errors can cause the same exit code. Fairseq does something odd to differentiate CUDA errors from the rest (https://github.com/pytorch/fairseq/blob/50a671f78d0c8de0392f924180db72ac9b41b801/fairseq/trainer.py#L283):

```
except RuntimeError as e:
  if 'out of memory' in str(e):
    # do something about CUDA out of memory error
  else:
    raise e
```

Reply:
Yeah, except we care mostly about "device-side asserts" (ie page faults) rather than CUDA OOM. I'm not sure what to look for in the error message.

Reply:
#174 has some examples of stack traces that we care about

Reply:
Yeah, so checking `if 'RuntimeError: CUDA error: device-side assert triggered' in str(e)` should do the trick, although not in a very clean way.

Reply:
Nice to see the RuntimeError log. I found that the long sequence in my test script actually produced another RuntimeError:
`RuntimeError: CUDA error: CUBLAS_STATUS_NOT_INITIALIZED when calling cublasCreate(handle)`, which is different from `device-side assert triggered`.

Reply:
We might want to catch "CUDA error" generically then.

Reply:
Addressed in the new update.

Reply:
I disabled the cuda tests in travis because travis is a cpu server.